drivers/led/lp50xx.c

```diff
@@ -0,0 +1,371 @@
+/*
+ * Copyright (c) 2020 Seagate Technology LLC
+ * Copyright (c) 2022 Grinn
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief LP50xx LED controller
+ */
+#include <errno.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/led.h>
+#include <zephyr/drivers/led/lp50xx.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+
+#define LOG_LEVEL CONFIG_LED_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(lp50xx);
+
+#include "led_context.h"
+
+/* Common registers */
+#define LP50XX_DEVICE_CONFIG0		0x00
+#define LP50XX_DEVICE_CONFIG1		0x01
+#define LP50XX_LED_CONFIG0		0x02
+
+/* LP5009/12 registers */
+#define LP5012_LED0_BRIGHTNESS		0x07
+#define LP5012_OUT0_COLOR		0x0b
+
+/* LP5018/24 registers */
+#define LP5024_LED0_BRIGHTNESS		0x07
+#define LP5024_OUT0_COLOR		0x0f
+
+/* LP5030/36 registers */
+#define LP5036_LED_CONFIG1		0x03
+#define LP5036_LED0_BRIGHTNESS		0x08
+#define LP5036_OUT0_COLOR		0x14
+
```

Comment:
`include/zephyr/drivers/led/lp50xx.h` is present because I have been struggling with the channel API which requires to expose some internal registers to make it work for this component (the base addresses of brightness and color registers). Which is why the info is duplicated and I would like to keep it that way because I have been thinking about a solution that would hide these registers and make the channel API more abstract.

I have been thinking of submitting an RFC (not done yet) to make the channel API more like the sensor interface using an enum as an argument to `led_write_channels()` to indicate which channels should be updated. I know that will make it less like the led_strip API but it should be more obvious for this kind of component (which seems to be the only one to use it ?)

Reply:
It is OK if you wan to keep this registers definitions duplicated. 

About the channel API, note that the [IS31FL3733 driver](https://github.com/zephyrproject-rtos/zephyr/pull/56673) is now using it.

One of the original intention of the channel API was to give users a direct access to the hardware channel registers. Also it was designed to be simple (at least for the driver) and compatible with the LED strip API. For my part I am not a big fan of the idea of abstracting the channels registers. If you want, I am open to discuss it on discord.

Reply:
OK, so to keep the current design, the `write_channels` API should use the LP50XX_LED_CONFIG0 register as base channel. This will make the bank feature work through raw accesses.
But for my part, I am not a big fan of having an API for raw access to a device as it may introduce deltas between configuration and register content... I think I will take the opportunity to join you on discord to discuss it.

Reply:
With the current implementation, the `LP50XX_LED_CONFIGx` registers are not exposed through the `write_channels` API. It starts with the `bank brightness` register. Please just keep it this way for this PR. And yes, we can discuss future changes for the `lp50xx` driver and the `write_channel` API.

Reply:
It is absolutely pointless to expose the `bank` registers without exposing the `led_config` registers. I would suggest to expose the `led_config` registers or offer a function in `lp50xx.h` that enables access to bank mode. Something like https://github.com/metratec/zephyr/blob/96c19193b00add6d939a4e0dad96179c7d1fcafe/include/zephyr/drivers/led/lp50xx.h#L71

Reply:
If you expose `lp50xx_config.chip` (and `enum lp50xx_chips`) you can define all the registers in `lp50xx.h`. The big advantage is, that the app doesn't need to know which flavour the LP50xx actually is.

See https://github.com/metratec/zephyr/blob/96c19193b00add6d939a4e0dad96179c7d1fcafe/include/zephyr/drivers/led/lp50xx.h#L23

With `LP5012_NUM_CHANNELS`, `LP5024_NUM_CHANNELS` and `LP5036_NUM_CHANNELS` it's still not obvious which one to pick for LP5009, LP5018 and LP5030.

Reply:
> It is absolutely pointless to expose the `bank` registers without exposing the `led_config` registers.

Yes it is. But the `write_channel` API is about giving raw access to channel registers but not to the LED configuration registers. More work is needed in the driver to allow bank configuration.

The point here is to let this PR just be about adding support for more controller models. This way it can finally be merged (after the failures of https://github.com/zephyrproject-rtos/zephyr/pull/46134 and https://github.com/zephyrproject-rtos/zephyr/pull/54594). Adding support for the hardware bank feature can come next with another PR. It will require further discussions about the `write_channels` API.